velox/exec/tests/SimpleAggregateAdapterTest.cpp

```diff
@@ -481,5 +483,234 @@ TEST_F(SimpleCountNullsAggregationTest, basic) {
   testAggregations({vectors}, {}, {"simple_count_nulls(c2)"}, {expected});
 }
 
+// A testing aggregation function that uses the function state.
+class FunctionStateTestAggregate {
+ public:
+  using InputType = Row<int64_t>; // Input vector type wrapped in Row.
+  using IntermediateType = int64_t; // Intermediate result type.
+  using OutputType = int64_t; // Output vector type.
```

Comment:
Thank you for adding the unit test!

Could we make the input, intermediate, and result types all be different, so that in initialize(), we can check that when step is partial, rawInputType is always the same as InputType, and when step is intermediate, rawInputType is always the same as IntermediateType? Also in initialize(), we can check that result type is expected.

Also, could we let this UDAF receive two arguments and make one of them constant literal in the aggregation function calls, e.g., call something like simple_function_state_agg(c0, 1) in unit tests. Then, let's check in initialize for the partial and main aggregation function that constantInputs is set correctly.

Reply:
Due to the fact that the step passed to `simple_function_state_agg_partial` for the function state is always partial, we cannot know whether the current step is the final step when checking `rawInputTypes`, `resultType` and `constantInputs`. Do you think we need to introduce a `companionStep` in function state?

Reply:
Optimizations have been made to the test cases.